File: src/main/java/compiler/tokenizer/Tokenizer.java
package compiler.tokenizer;

import java.util.ArrayList;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class Tokenizer {

	private ArrayList<TokenData> tokenDatas;
	
	private String str;
	
	private Token lastToken;
	private boolean pushBack;
	
	public Tokenizer(String str) {
		this.tokenDatas = new ArrayList<TokenData>();
		this.str = str;
		

		tokenDatas.add(new TokenData(Pattern.compile("^((-)?[0-9]+)"), TokenType.INTEGER_LITERAL));
		tokenDatas.add(new TokenData(Pattern.compile("^((-)?[0-9]+[.][0-9])"), TokenType.DOUBLE_LITERAL));
		tokenDatas.add(new TokenData(Pattern.compile("^([+][=])"), TokenType.ADD_OPERATOR));
		tokenDatas.add(new TokenData(Pattern.compile("^([-][=])"), TokenType.SUBTRACT_OPERATOR));
		tokenDatas.add(new TokenData(Pattern.compile("^([*][=])"), TokenType.MULTIPLY_OPERATOR));
		tokenDatas.add(new TokenData(Pattern.compile("^([/][=])"), TokenType.DIVIDE_OPERATOR));
		tokenDatas.add(new TokenData(Pattern.compile("^(\".*\")"), TokenType.STRING_LITERAL));
		tokenDatas.add(new TokenData(Pattern.compile("^([;])"), TokenType.END_STATEMENT));
		tokenDatas.add(new TokenData(Pattern.compile("^([:])"), TokenType.COLON));
		tokenDatas.add(new TokenData(Pattern.compile("^([==])"), TokenType.EQUAL_TO));
		tokenDatas.add(new TokenData(Pattern.compile("^([<])"), TokenType.SMALLER_THAN));
		tokenDatas.add(new TokenData(Pattern.compile("^([<=])"), TokenType.SMALLER_THAN_EQUAL));
		tokenDatas.add(new TokenData(Pattern.compile("^([>])"), TokenType.LARGER_THAN));
		tokenDatas.add(new TokenData(Pattern.compile("^([>=])"), TokenType.LARGER_THAN_EQUAL));
		tokenDatas.add(new TokenData(Pattern.compile("^([a-zA-Z][a-zA-Z0-9]*)"), TokenType.IDENTIFIER));


		for (String t : new String[] { "=", "\\(", "\\)", "\\.", "\\,", "\\'" }) {
			tokenDatas.add(new TokenData(Pattern.compile("^(" + t + ")"), TokenType.TOKEN));
		}
	}
	
	public Token nextToken() {
		str = str.trim();
		
		if (pushBack) {
			pushBack = false;
			return lastToken;
		}
		
		if (str.isEmpty()) {
			return (lastToken = new Token("", TokenType.EMPTY));
		}
		
		for (TokenData data : tokenDatas) {
			Matcher matcher = data.getPattern().matcher(str);
			
			if (matcher.find()) {
				String token = matcher.group().trim();
				str = matcher.replaceFirst("");
				
				if (data.getType() == TokenType.STRING_LITERAL) {
					return (lastToken = new Token(token.substring(1, token.length() - 1), TokenType.STRING_LITERAL));
				}
				
				else {
					return (lastToken = new Token(token, data.getType()));
				}
			}
		}
		
		throw new IllegalStateException("Could not parse " + str);
	}
	
	public boolean hasNextToken() {
		return !str.isEmpty();
	}
	
	public void pushBack() {
		if (lastToken != null) {
			this.pushBack = true;
		}
	}
}

File: src/main/scala/compiler/tokenizer/Tokenizer.scala
/*
 * Cobalt Programming Language Compiler
 * Copyright (C) 2017  Cobalt
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package compiler.tokenizer


import compiler.tokenizer.tokens._
import compiler.tokenizer.tokens.constants._
import compiler.tokenizer.tokens.keywords.modifiers._
import compiler.tokenizer.tokens.keywords.{ExtendsToken, ImplementsToken}
import compiler.tokenizer.tokens.operators._

import scala.collection.mutable.ListBuffer

class Tokenizer(var line: String) {

  private val tokenDatas: ListBuffer[TokenType] = ListBuffer[TokenType](

    /* Modifiers */
    new PublicToken,
    new ProtectedToken,
    new InternalToken,
    new AbstractToken,
    new OverrideToken,
    new OpenToken,

    // keywords
    new ExtendsToken,
    new ImplementsToken,

    new ReturnTypeToken,

    new FloatLiteralToken,
    new DoubleLiteralToken,
    new ByteLiteralToken,
    new ShortLiteralToken,
    new LongLiteralToken,
    new IntegerLiteralToken,

    new StringLiteralToken,
    new CharacterLiteralToken,

    new AddOperatorToken,
    new SubtractOperatorToken,
    new MultiplyOperatorToken,
    new DivideOperatorToken,
    new ModulusOperatorToken,

    new EndStatementToken,
    new ColonToken,
    new UnderscoreToken,

    new EqualToToken,
    new SmallerThanToken,
    new SmallerThanEqualToken,
    new LargerThanToken,
    new LargerThanEqualToken,

    new IdentifierToken,

    new AssignmentToken,
    new OpeningBracketToken,
    new ClosingBracketToken,
    new FullStopToken,
    new CommaToken,
    new ApostropheToken
  )

  //line = StringEscapeUtils.escapeJava(line)

  /**
    * Returns the next token without removing it from the line
    *
    * @return The next token
    */
  def peek: Token = {
    line = line.trim

    if (line.isEmpty) {
      new Token("", new EmptyToken)
    } else {
      for (data <- tokenDatas) {
        val matched = data.getRegex().findFirstIn(line)
        if (matched.isDefined && matched.get == line.substring(0, matched.get.size)) {
          val token: String = matched.getOrElse("")
          return new Token(token, data)
        }
      }
      throw new IllegalStateException("Could not parse:" + line)
    }
  }

  /**
    * Returns the next token and removes it from the line
    *
    * @return The next token
    */
  def nextToken: Token = {
    line = line.trim

    if (line.isEmpty) {
      new Token("", new EmptyToken)
    } else {
      for (data <- tokenDatas) {
        val matched = data.getRegex().findFirstIn(line)
        if (matched.isDefined && matched.get == line.substring(0, matched.get.size)) {
          val token: String = matched.getOrElse("")
          line = data.getRegex().replaceFirstIn(line, "")
          return new Token(token, data)
        }
      }
      throw new IllegalStateException("Could not parse:" + line)
    }
  }



  /**
    * Returns whether the next token exists.
    *
    * @return
    */
  def hasNextToken: Boolean = !line.isEmpty
}
